apps/examples/hello/hello_main.c

```diff
@@ -68,5 +70,46 @@ int hello_main(int argc, char *argv[])
 #endif
 {
 	printf("Hello, World!!\n");
+	/* Call stay IOCTL to keep PM domain in current state because maybe we know we need to do some
+	 * non-frequent work with time gaps and dont want the board to sleep in between
+	 */
+	pm_ioctl(PM_IOC_STAY);
+
+	{
+		//The work happens
+	}
+
+	/* The work is done, now relax the stay on the current state of the domain */
+	pm_ioctl(PM_IOC_RELAX);
+
+	{
+		//App does some other non-blocking work
+	}
+
+	/* Now, maybe the app has completed the major part of its work and is suggesting that the board
+	 * can go to sleep, given that there is nothing else going on.
+	 * This does not mean that the board will immediately be put to sleep. If there is a stay applied
+	 * by another application, then this SLEEP request will be DISCARDED(can be discussed if otherwise).
+	 */
+	pm_ioctl(PM_IOC_SLEEP);
+
+	/* However, as discussed before, we may need the application to request/suggest the core to go to
```

Comment:
This could also be handled in PM_IOCTL, but we might have to revised the input parameter for PM_IOCTL, to pass in the sleep duration from application layer.
Note that, if there is other interrupt source coming in before the timer ends, the core will be awake immediately, and the timer will be cancelled when the time reaches.

Reply:
1. Yes I agree, IOCTL function definition needs to be revised.
2. Third parameter - suppose App requests for 5 seconds sleep NOW @ time = t1. Requests goes through IOCTL to pm module. Some checks are done to determine the feasibility of sleep, etc.
Then board id ready to go to sleep and a timer interrupt needs to be set.
At this time, t2 = g_system_timer
t2 > t1
Hence, timer interrupt should be set for **5 seconds - (t2 - t1)**
This is similar to adjusting the delay in pm_timer callback function.

Reply:
The enum in pm.h you provided has 3 meaningful states which are PM_IOC_STAY/RELAX/SLEEP, but looks like there is one more missing out, which is PM_IOC_TIMEDSLEEP, could you clarify on this?

Reply:
> 1. Yes I agree, IOCTL function definition needs to be revised.
> 2. Third parameter - suppose App requests for 5 seconds sleep NOW @ time = t1. Requests goes through IOCTL to pm module. Some checks are done to determine the feasibility of sleep, etc.
>    Then board id ready to go to sleep and a timer interrupt needs to be set.
>    At this time, t2 = g_system_timer
>    t2 > t1
>    Hence, timer interrupt should be set for **5 seconds - (t2 - t1)**
>    This is similar to adjusting the delay in pm_timer callback function.

Sorry, I still couldn't quite get the meaning here, even if we need to get the current time(t1), I think g_system_timer will be invoke similarly as getting t2. Additionally, does the difference between t2 and t1 includes the timer gap between each state transition? I believe ready to go to sleep means it will be already in PM_SLEEP state, so if the trigger point starts from PM_NORMAL, it will go through 3 state transitions before reaching PM_SLEEP, hence if the timer gap between each state is considered, there might be (t2 - t1) > 5seconds (sleeping interval), since we all knew no negative input could be provided for timer, what is the consideration for such cases?

Reply:
Yes the enum has TIMEDSLEEP missing.
This commit is just to get things started. It will need changes and refinement. I created it so we can discuss on the basis of these ideas. Similarly, ioctl does not accept more than one argument currently, that may also need changes.

Reply:
Understood, I think for the TIMEDSLEEP enum, either it is following ladder like operation, or it will be directly going to sleep, it should be able to combine with RELAX or SLEEP enum into one, because the difference is just it will be utilising timer as one of the potential interrupt source. 
In conclusion, I think 3 enum will be enough for your operation. How do you think?

Reply:
If we can work with just 3, that is okay. If we need more, that is also fine.
I pushed these files to show sample structure. Specific details may change based on implementation.

Reply:
regarding the question about timer adjustment, t1 and t2 in the comment above
Please take a look a this: -https://github.com/ziyik/TizenRT/blob/4efc0f42b5c9829fe5c2d502a27a330dabd5435b/os/pm/pm_changestate.c#L117-L124

We need to set the timer for time = (requested_time - time_elapsed_since_request)

We may be able to do this with the 3 parameters that I suggested, or with some other implementation.
If you have a different idea than using the 3 parameters that I mentioned, please suggest.

Reply:
Understood, let's continue the conversation after the first version of implementation is released. I will update you again.

Thank you~